### PR TITLE
Added cli capabilities for dx paths

### PR DESCRIPTION
**Pull request by anujkumar93**
Changed list to call walkfiles for dx paths.
Fixed cli bugs for dx paths related to 'cd' and 'pwd'

### CODE REVIEW

File: stor/cli.py [hunk 1]
```diff
@@ -415,6 +423,8 @@ def process_args(args):
         for key, val in args_copy.items() if val
     }
     try:
+        if func == stor.list and utils.is_dx_path(pth):
+            func = stor.walkfiles
```
Comment (jtratner) [inline stor/cli.py]: I think it makes more sense to have a wrapped list that looks like:

```
def _wrapped_list(path, **kwargs):
    """Use iterative walkfiles for DX paths, rather than trying to generate full list first"""
    if utils.is_dx_path(path):
        func = stor.walkfiles
    else:
        func = stor.list
    return func(path, **kwargs)
```

File: stor/cli.py [hunk 2]
```diff
@@ -186,10 +189,13 @@ def _env_chdir(pth):
     """Sets the new current working directory."""
     parser = _get_env()
     if utils.is_obs_path(pth):
-        service = Path(pth).drive.rstrip(':/')
```
Comment (jtratner) [inline stor/cli.py:189]: why did you need to make these changes? Did some test cases fail to motivate this?

File: stor/cli.py [hunk 3]
```diff
@@ -186,10 +189,13 @@ def _env_chdir(pth):
...
+        if pth == 'dx://':
+            service = 'dx'
+        else:
+            service = Path(pth).drive.rstrip(':/')
     else:
         raise ValueError('%s is an invalid path' % pth)
-    if pth != Path(pth).drive:
+    if pth != 'dx://' and pth != Path(pth).drive:
```
Comment (jtratner) [inline stor/cli.py:198]: huh?

File: stor/cli.py [hunk 4]
```diff
@@ -271,7 +277,7 @@ def get_path(pth, mode=None):
 
 def _to_url(path):
     if stor.is_filesystem_path(path):
-        raise ValueError('must be swift or s3 path')
+        raise ValueError('must be swift or s3 or dx path')
```
Comment (jtratner) [inline stor/cli.py]: will dx paths actually work for this method?

**Review by jtratner — COMMENTED** [review]
looks like right path, a couple changes + explanation to make (+ request for tests to demonstrate bugs).

Also: after looking at the code, do you think s3 and swift could also use walkfiles?

**Comment by jtratner** [comment]
actually since `walkfiles()` is defined for `swift` and `s3` as calling `list()` under the head, may as well switch 'em all to `list()`

**Comment by jtratner** [comment]
you'd need to wrap a bit - i.e., as:

```
if args.starts_with:
    pattern = '%s*' % args.starts_with
else:
    pattern = args.pattern
for i, elem in enumerate(stor.walkfiles(path, pattern=pattern), 1):
    yield elem
    if i >= limit:
        break
```

File: stor/cli.py (hunk 2 above)
Comment (anujkumar93) [reply stor/cli.py:189]: DXPath is implemented to always require a project. So, Path('dx://') will fail in general.

File: stor/cli.py (hunk 3 above)
Comment (anujkumar93) [reply stor/cli.py:198]: Same reason above.

File: stor/cli.py (hunk 4 above)
Comment (anujkumar93) [reply stor/cli.py]: My bad, to_url is not implemented for dx paths. Removed.

**Comment by anujkumar93** [comment]
I thought the point of moving list to walkfiles was to prevent loading the whole list into memory. We can't avoid that with swift and s3 (because walkfiles calls list), but we can with dx

**Comment by anujkumar93** [comment]
> you'd need to wrap a bit - i.e., as:
> 
> ```
> if args.starts_with:
>     pattern = '%s*' % args.starts_with
> else:
>     pattern = args.pattern
> for i, elem in enumerate(stor.walkfiles(path, pattern=pattern), 1):
>     yield elem
>     if i >= limit:
>         break
> ```

What's this for? List takes in starts_with and limit arguments, and starts_with is the subfolder you want to search in, doesn't take glob patterns. Walkfiles takes in a pattern, which should be glob, I think you meant to wrap around that.

**Comment by jtratner** [comment]
my bad - I should have written `let's switch them all to walkfiles`

```
def walkfiles_as_list(path, starts_with, limit):
    if args.starts_with:
        pattern = '%s*' % args.starts_with
    else:
        pattern = None
    for i, elem in enumerate(stor.walkfiles(path, pattern=pattern), 1):
        yield elem
        if i >= limit:
            break
```

(the limit part is just to ensure that, if specified, we respect it)

**Comment by anujkumar93** [comment]
Okay I get what you're trying to say. why make `starts_with` into glob though? starts_with (at least for dx) is a path to a subdir, not a glob pattern match against the file name

**Review by jtratner — APPROVED** [review]
cool lgtm - let's merge this
